File: packages/react-native-tracer-provider/ios/ReactNativeTracerProvider/ReactNativeTracerProviderModule.swift
import ObjectiveC
import Foundation
import OpenTelemetryApi
import EmbraceIO
import os

/*
 NOTE: There's currently a bit of duplication between this and code in packages/core, particularly https://github.com/embrace-io/embrace-react-native-sdk/blob/7c54b59362adfc93f7f997db89db179950a50e8b/packages/core/ios/RNEmbrace/SpanRepository.swift
 
 The idea will be to have this package power all span features in 6.0 and remove that code from
 the core package so living with the duplication for now
 */

private typealias OpenTelemetryAttributes = [String: OpenTelemetryApi.AttributeValue]

private let LINK_ATTRIBUTES_KEY = "attributes"
private let LINK_SPAN_CONTEXT_KEY = "context"
private let SPAN_CONTEXT_TRACE_ID_KEY = "traceId"
private let SPAN_CONTEXT_SPAN_ID_KEY = "spanId"
private let SPAN_STATUS_CODE_KEY = "code"
private let SPAN_STATUS_MESSAGE_KEY = "message"

// Should not get hit under normal circumstances, add as a guard against misinstrumentation
private let MAX_STORED_SPANS = 10000

@objc(ReactNativeTracerProviderModule)
class ReactNativeTracerProviderModule: NSObject {
  private let tracersQueue = DispatchQueue(
    label: "io.embrace.reactnativetracerprovider.tracers",
    attributes: .concurrent
  )
  private let activeSpansQueue = DispatchQueue(
    label: "io.embrace.reactnativetracerprovider.activeSpans",
    attributes: .concurrent
  )
  private let completedSpansQueue = DispatchQueue(
    label: "io.embrace.reactnativetracerprovider.completedSpans",
    attributes: .concurrent
  )
  private var tracers = [String: Tracer]()
  private var activeSpans = [String: Span]()
  private var completedSpans = [String: Span]()
  private var tracerProvider: TracerProvider!
  private var log = OSLog(subsystem: "Embrace", category: "ReactNativeTracerProviderModule")

  /**
   * Various deserializer helpers to go to and from NSDictionary / NSArray to
   * actual OTEL API objects
   */
  private func attributesFrom(dict: NSDictionary) -> OpenTelemetryAttributes {
    var attributes = OpenTelemetryAttributes()

    for (key, value) in dict {
      if let key = key as? String {
        if let value = value as? String {
          attributes.updateValue(AttributeValue(value), forKey: key)
        } else if let value = value as? Bool {
          attributes.updateValue(AttributeValue(value), forKey: key)
        } else if let value = value as? Int {
          attributes.updateValue(AttributeValue(value), forKey: key)
        } else if let value = value as? Double {
          attributes.updateValue(AttributeValue(value), forKey: key)
        } else if let value = value as? [String] {
          attributes.updateValue(AttributeValue(value), forKey: key)
        } else if let value = value as? [Bool] {
          attributes.updateValue(AttributeValue(value), forKey: key)
        } else if let value = value as? [Int] {
          attributes.updateValue(AttributeValue(value), forKey: key)
        } else if let value = value as? [Double] {
          attributes.updateValue(AttributeValue(value), forKey: key)
        } else {
          os_log("invalid attribute for key:%@", log: log, type: .error, key)
          continue
        }
      } else {
        os_log("invalid attribute key", log: log, type: .error)
        continue
      }
    }

    return attributes
  }

  private func spanContextFrom(dict: NSDictionary) -> SpanContext? {
    let traceId = dict.value(forKey: SPAN_CONTEXT_TRACE_ID_KEY)
    let spanId = dict.value(forKey: SPAN_CONTEXT_SPAN_ID_KEY)

    if let traceId = traceId as? String, let spanId = spanId as? String {
      return SpanContext.create(
        traceId: TraceId(fromHexString: traceId),
        spanId: SpanId(fromHexString: spanId),
        traceFlags: TraceFlags(),
        traceState: TraceState()
      )
    } else {
      return nil
    }
  }

  private func spanContextToDict(spanContext: SpanContext) -> NSDictionary {
    return NSDictionary(dictionary: [
      SPAN_CONTEXT_TRACE_ID_KEY: spanContext.traceId.hexString,
      SPAN_CONTEXT_SPAN_ID_KEY: spanContext.spanId.hexString
    ])
  }

  private func dateFrom(ms: Double) -> Date {
    return Date(timeIntervalSince1970: TimeInterval(ms / 1000.0))
  }

  private func getTracerKey(name: String, version: String, schemaUrl: String) -> String {
    return "\(name) \(version) \(schemaUrl)"
  }

  private func getSpan(spanBridgeId: String) -> Span? {
    var span: Span?

    activeSpansQueue.sync {
      span = activeSpans[spanBridgeId]
    }

    if span == nil {
        completedSpansQueue.sync {
            span = completedSpans[spanBridgeId]
        }
    }

    if span == nil {
      os_log("could not retrieve span with id: %@", log: log, type: .error, spanBridgeId)
    }

    return span
  }

  /**
   * Methods to allow the JS side to conform to @opentelemetry-js/api
   */

  @objc(setupTracer:version:schemaUrl:)
  func setupTracer(name: String, version: String, schemaUrl: String) {
    if tracerProvider == nil {
      if let isStarted = Embrace.client?.started {
        tracerProvider = OpenTelemetry.instance.tracerProvider
      } else {
        os_log("cannot access tracer provider, Embrace SDK has not been started", log: log, type: .error)
        return
      }
    }

    let id = getTracerKey(name: name, version: version, schemaUrl: schemaUrl)
    var existingTracer: Tracer?

    tracersQueue.sync {
      existingTracer = tracers[id]
    }

    if existingTracer != nil {
      return
    }

    // No current way to set schemaURL in the OTEL Swift API, ignoring
    let tracer = tracerProvider.get(instrumentationName: name, instrumentationVersion: version)

    tracersQueue.async(flags: .barrier) {
      self.tracers.updateValue(tracer, forKey: id)
    }
  }

  @objc(startSpan:tracerVersion:tracerSchemaUrl:spanBridgeId:name:kind:time:attributes:links:parentId:resolve:reject:)
  func startSpan(tracerName: String, tracerVersion: String, tracerSchemaUrl: String,
                 spanBridgeId: String, name: String, kind: String, time: Double,
                 attributes: NSDictionary, links: NSArray, parentId: String,
                 resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    let tracerKey = getTracerKey(name: tracerName, version: tracerVersion, schemaUrl: tracerSchemaUrl)
    var tracer: Tracer?

    tracersQueue.sync {
      tracer = tracers[tracerKey]
    }

    if tracer == nil {
      reject("START_SPAN", "tracer not found", nil)
      return
    }

    let spanBuilder = tracer!.spanBuilder(spanName: name)

    // Set kind
    if !kind.isEmpty {
      if let spanKind = SpanKind(rawValue: kind.lowercased()) {
        spanBuilder.setSpanKind(spanKind: spanKind)
      } else {
        os_log("invalid kind for startSpan: %@", log: log, type: .error, kind)
      }
    }

    // Set time
    if !time.isZero {
      spanBuilder.setStartTime(time: dateFrom(ms: time))
    }

    // Set attributes
    for (key, value) in attributesFrom(dict: attributes) {
      spanBuilder.setAttribute(key: key, value: value)
    }

    // Set links
    for link in links {
      if let link = link as? NSDictionary {
        let linkSpanContext = link.value(forKey: LINK_SPAN_CONTEXT_KEY)
        let linkAttributes = link.value(forKey: LINK_ATTRIBUTES_KEY)

        if let linkSpanContext = linkSpanContext as? NSDictionary,
           let spanContext = spanContextFrom(dict: linkSpanContext) {
          if linkAttributes != nil, let linkAttributes = linkAttributes as? NSDictionary, linkAttributes.count > 0 {
            spanBuilder.addLink(spanContext: spanContext, attributes: attributesFrom(dict: linkAttributes))
          } else {
            spanBuilder.addLink(spanContext: spanContext)
          }
        } else {
          os_log("invalid span context for link", log: log, type: .error)
          continue
        }
      }
    }

    // Set parent
    if !parentId.isEmpty, let parent = getSpan(spanBridgeId: parentId) {
      spanBuilder.setParent(parent)
    } else {
      spanBuilder.setNoParent()
    }

    let span = spanBuilder.startSpan()
    activeSpansQueue.async(flags: .barrier) {
        if self.activeSpans.count > MAX_STORED_SPANS {
          os_log("too many active spans being tracked, ignoring", log: self.log, type: .error)
          reject("START_SPAN", "too many active spans being tracked, ignoring", nil)
        } else {
          self.activeSpans.updateValue(span, forKey: spanBridgeId)
          resolve(self.spanContextToDict(spanContext: span.context))
        }
    }
  }

  @objc(setAttributes:attributes:)
  func setAttributes(spanBridgeId: String, attributes: NSDictionary) {
    if let span = getSpan(spanBridgeId: spanBridgeId) {
      for (key, value) in attributesFrom(dict: attributes) {
        span.setAttribute(key: key, value: value)
      }
    }
  }

  @objc(addEvent:eventName:attributes:time:)
  func addEvent(spanBridgeId: String, eventName: String, attributes: NSDictionary, time: Double) {
    if let span = getSpan(spanBridgeId: spanBridgeId) {
      if time.isZero {
        span.addEvent(name: eventName, attributes: attributesFrom(dict: attributes))
      } else {
        span.addEvent(name: eventName, attributes: attributesFrom(dict: attributes), timestamp: dateFrom(ms: time))
      }
    }
  }

  @objc(addLinks:links:)
  func addLinks(spanBridgeId: String, links: NSArray) {
    // not supported by the OTEL Swift API currently
    os_log("adding links is not currently supported", log: log, type: .error)
  }

  @objc(setStatus:status:)
  func setStatus(spanBridgeId: String, status: NSDictionary) {
    if let span = getSpan(spanBridgeId: spanBridgeId),
       let code = status.value(forKey: SPAN_STATUS_CODE_KEY),
       let code = code as? String {
      switch code {
      case "UNSET":
        span.status = .unset
      case "OK":
        span.status = .ok
      case "ERROR":
        if let message = status.value(forKey: SPAN_STATUS_MESSAGE_KEY), let message = message as? String {
          span.status = .error(description: message)
        } else {
          span.status = .error(description: "")
        }
      default:
        os_log("invalid status for setStatus: %@", log: log, type: .error, code)
      }
    }
  }

  @objc(updateName:name:)
  func updateName(spanBridgeId: String, name: String) {
    if let span = getSpan(spanBridgeId: spanBridgeId) {
      span.name = name
    }
  }

  @objc(endSpan:time:)
  func endSpan(spanBridgeId: String, time: Double) {
    if let span = getSpan(spanBridgeId: spanBridgeId) {
      if time.isZero {
        span.end()
      } else {
        span.end(time: dateFrom(ms: time))
      }

      activeSpansQueue.async(flags: .barrier) {
        self.activeSpans.removeValue(forKey: spanBridgeId)
      }

      completedSpansQueue.async(flags: .barrier) {
        if self.completedSpans.count > MAX_STORED_SPANS {
          os_log("too many completed spans being tracked, ignoring", log: self.log, type: .error)
          return
        }

        self.completedSpans.updateValue(span, forKey: spanBridgeId)
      }
    }
  }

  @objc(clearCompletedSpans)
  func clearCompletedSpans() {
    completedSpansQueue.async(flags: .barrier) {
      self.completedSpans.removeAll()
    }
  }
 }

File: packages/react-native-tracer-provider/android/src/main/java/io/embrace/reactnativetracerprovider/ReactNativeTracerProviderModule.kt
package io.embrace.reactnativetracerprovider

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeMap
import io.embrace.android.embracesdk.Embrace
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.common.Attributes
import io.opentelemetry.api.trace.Span
import io.opentelemetry.api.trace.SpanContext
import io.opentelemetry.api.trace.SpanKind
import io.opentelemetry.api.trace.StatusCode
import io.opentelemetry.api.trace.TraceFlags
import io.opentelemetry.api.trace.TraceState
import io.opentelemetry.api.trace.Tracer
import io.opentelemetry.api.trace.TracerProvider
import io.opentelemetry.context.Context
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.logging.Logger

/**
 * Needed so WritableNativeMap can be swapped out for JavaOnlyMap during testing as the former
 * breaks in a mocked React Native environment
 */
interface WritableMapBuilder {
    fun build(): WritableMap
}

class WritableNativeMapBuilder : WritableMapBuilder {
    override fun build(): WritableMap {
        return WritableNativeMap()
    }
}

private const val LINK_ATTRIBUTES_KEY = "attributes"
private const val LINK_SPAN_CONTEXT_KEY = "context"
private const val SPAN_CONTEXT_TRACE_ID_KEY = "traceId"
private const val SPAN_CONTEXT_SPAN_ID_KEY = "spanId"
private const val SPAN_STATUS_CODE_KEY = "code"
private const val SPAN_STATUS_MESSAGE_KEY = "message"

// Should not get hit under normal circumstances, add as a guard against misinstrumentation
private const val MAX_STORED_SPANS = 10000

class ReactNativeTracerProviderModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private val log = Logger.getLogger("[Embrace]")
    private val tracers = ConcurrentHashMap<String, Tracer>()
    private val activeSpans = ConcurrentHashMap<String, Span>()
    private val completedSpans = ConcurrentHashMap<String, Span>()
    private var tracerProvider: TracerProvider? = null
    private var writableMapBuilder: WritableMapBuilder

    override fun getName() = "ReactNativeTracerProviderModule"

    /**
     * Various deserializer helpers to go to and from the bridge Readable/Writable Array/Maps to
     * actual OTEL API objects
     */

    private fun stringListFromReadableArray(array: ReadableArray): List<String> {
        val list: MutableList<String> = mutableListOf()
        for (i in 0..array.size() - 1) {
            list.add(array.getString(i))
        }
        return list
    }

    private fun doubleListFromReadableArray(array: ReadableArray): List<Double> {
        val list: MutableList<Double> = mutableListOf()
        for (i in 0..array.size() - 1) {
            list.add(array.getDouble(i))
        }
        return list
    }

    private fun booleanListFromReadableArray(array: ReadableArray): List<Boolean> {
        val list: MutableList<Boolean> = mutableListOf()
        for (i in 0..array.size() - 1) {
            list.add(array.getBoolean(i))
        }
        return list
    }

    private fun attributesFromReadableMap(map: ReadableMap): Attributes {
        val it = map.keySetIterator()
        val builder = Attributes.builder()

        while (it.hasNextKey()) {
            val key = it.nextKey()
            val type = map.getType(key)

            when (type) {
                ReadableType.Boolean -> builder.put(key, map.getBoolean(key))
                ReadableType.Number -> builder.put(key, map.getDouble(key))
                ReadableType.String -> builder.put(key, map.getString(key) ?: "")
                ReadableType.Array -> {
                    val array = map.getArray(key)
                    val arrayType = array?.getType(0)
                    when (arrayType) {
                        ReadableType.Boolean -> builder.put(AttributeKey.booleanArrayKey(key), booleanListFromReadableArray(array))
                        ReadableType.Number -> builder.put(AttributeKey.doubleArrayKey(key), doubleListFromReadableArray(array))
                        ReadableType.String -> builder.put(AttributeKey.stringArrayKey(key), stringListFromReadableArray(array))
                        else -> {
                            log.warning("invalid attribute key: $key")
                        }
                    }
                }
                else -> {
                    log.warning("invalid attribute value for key: $key")
                }
            }
        }

        return builder.build()
    }

    private fun spanContextFromReadableMap(map: ReadableMap): SpanContext {
        val traceId = map.getString(SPAN_CONTEXT_TRACE_ID_KEY) ?: ""
        val spanId = map.getString(SPAN_CONTEXT_SPAN_ID_KEY) ?: ""

        val spanContext = SpanContext.create(
            traceId,
            spanId,
            TraceFlags.getDefault(),
            TraceState.getDefault()
        )

        return spanContext
    }

    private fun spanContextToWritableMap(spanContext: SpanContext?): WritableMap {
        val map = this.writableMapBuilder.build()

        map.putString(SPAN_CONTEXT_TRACE_ID_KEY, spanContext?.traceId ?: "")
        map.putString(SPAN_CONTEXT_SPAN_ID_KEY, spanContext?.spanId ?: "")

        return map
    }

    init {
        writableMapBuilder = WritableNativeMapBuilder()
    }

    /**
     * Exposed for unit testing to allow a writableMapBuilder to be injected
     */
    constructor(
        reactContext: ReactApplicationContext,
        writableMapBuilder: WritableMapBuilder
    ) : this(reactContext) {
        this.writableMapBuilder = writableMapBuilder
    }

    /**
     * Exposed for unit testing to allow a different trace provider and writableMapBuilder to be injected
     */
    constructor(
        reactContext: ReactApplicationContext,
        tracerProvider: TracerProvider,
        writableMapBuilder: WritableMapBuilder
    ) : this(reactContext) {
        this.tracerProvider = tracerProvider
        this.writableMapBuilder = writableMapBuilder
    }

    private fun getTracerKey(name: String, version: String, schemaUrl: String): String {
        return "$name $version $schemaUrl"
    }

    private fun getSpan(spanBridgeId: String): Span? {
        val span = activeSpans[spanBridgeId] ?: completedSpans[spanBridgeId]
        if (span == null) {
            log.warning("could not retrieve span with bridge id: $spanBridgeId")
        }
        return span
    }

    /**
     * Methods to allow the JS side to conform to @opentelemetry-js/api
     */

    @ReactMethod
    fun setupTracer(name: String, version: String, schemaUrl: String) {
        if (tracerProvider == null) {
            if (!Embrace.getInstance().isStarted) {
                log.warning("cannot access tracer provider, Embrace SDK has not been started")
                return
            }

            tracerProvider = Embrace.getInstance().getOpenTelemetry().tracerProvider
        }

        val id = getTracerKey(name, version, schemaUrl)

        if (tracers.containsKey(id)) {
            // tracer is already setup
            return
        }

        val builder = tracerProvider?.tracerBuilder(name)

        if (version != "") {
            builder?.setInstrumentationVersion(version)
        }

        if (schemaUrl != "") {
            builder?.setSchemaUrl(schemaUrl)
        }

        val tracer = builder?.build()

        if (tracer != null) {
            tracers[id] = tracer
        }
    }

    @Suppress("LongParameterList")
    @ReactMethod
    fun startSpan(
        tracerName: String,
        tracerVersion: String,
        tracerSchemaUrl: String,
        spanBridgeId: String,
        name: String,
        kind: String,
        time: Double,
        attributes: ReadableMap,
        links: ReadableArray,
        parentId: String,
        promise: Promise
    ) {
        val tracer = tracers[getTracerKey(tracerName, tracerVersion, tracerSchemaUrl)]

        if (tracer == null) {
            promise.reject("START_SPAN", "tracer not found")
            return
        }

        val spanBuilder = tracer.spanBuilder(name)

        // Set kind
        if (kind.isNotEmpty()) {
            try {
                spanBuilder.setSpanKind(SpanKind.valueOf(kind))
            } catch (e: IllegalArgumentException) {
                log.warning("invalid kind for startSpan: $kind")
            }
        }

        // Set time
        if (time != 0.0) {
            spanBuilder.setStartTimestamp(time.toLong(), TimeUnit.MILLISECONDS)
        }

        // Set attributes
        spanBuilder.setAllAttributes(attributesFromReadableMap(attributes))

        // Set links
        for (i in 0..links.size() - 1) {
            val link = links.getMap(i)
            val linkSpanContext = link.getMap(LINK_SPAN_CONTEXT_KEY) ?: continue
            val linkAttributes = link.getMap(LINK_ATTRIBUTES_KEY)

            if (linkAttributes == null) {
                val spanContext = spanContextFromReadableMap(linkSpanContext)
                spanBuilder.addLink(spanContext)
            } else {
                spanBuilder.addLink(spanContextFromReadableMap(linkSpanContext), attributesFromReadableMap(linkAttributes))
            }
        }

        // Set parent
        if (parentId.isEmpty()) {
            spanBuilder.setNoParent()
        } else {
            val parent = getSpan(parentId)
            if (parent != null) {
                spanBuilder.setParent(parent.storeInContext(Context.root()))
            }
        }

        if (activeSpans.size > MAX_STORED_SPANS) {
            val msg = "too many active spans being tracked, ignoring"
            log.warning(msg)
            promise.reject("START_SPAN", msg)
        } else {
            val span = spanBuilder.startSpan()
            activeSpans[spanBridgeId] = span
            promise.resolve(spanContextToWritableMap(span.spanContext))
        }
    }

    @ReactMethod
    fun setAttributes(spanBridgeId: String, attributes: ReadableMap) {
        val span = getSpan(spanBridgeId) ?: return
        span.setAllAttributes(attributesFromReadableMap(attributes))
    }

    @ReactMethod
    fun addEvent(spanBridgeId: String, eventName: String, attributes: ReadableMap, time: Double) {
        val span = getSpan(spanBridgeId) ?: return

        if (time != 0.0) {
            span.addEvent(eventName, attributesFromReadableMap(attributes), time.toLong(), TimeUnit.MILLISECONDS)
        } else {
            span.addEvent(eventName, attributesFromReadableMap(attributes))
        }
    }

    @ReactMethod
    fun addLinks(spanBridgeId: String, links: ReadableArray) {
        val span = getSpan(spanBridgeId) ?: return
        for (i in 0..links.size() - 1) {
            val link = links.getMap(i)
            val linkSpanContext = link.getMap(LINK_SPAN_CONTEXT_KEY) ?: continue
            val linkAttributes = link.getMap(LINK_ATTRIBUTES_KEY)

            if (linkAttributes == null) {
                span.addLink(spanContextFromReadableMap(linkSpanContext))
            } else {
                span.addLink(spanContextFromReadableMap(linkSpanContext), attributesFromReadableMap(linkAttributes))
            }
        }
    }

    @ReactMethod
    fun setStatus(spanBridgeId: String, status: ReadableMap) {
        val span = getSpan(spanBridgeId) ?: return
        val statusCode = status.getString(SPAN_STATUS_CODE_KEY) ?: return
        val message = status.getString(SPAN_STATUS_MESSAGE_KEY) ?: ""

        try {
            if (message.isEmpty()) {
                span.setStatus(StatusCode.valueOf(statusCode))
            } else {
                span.setStatus(StatusCode.valueOf(statusCode), message)
            }
        } catch (e: IllegalArgumentException) {
            log.warning("invalid statusCode for setStatus: $statusCode")
        }
    }

    @ReactMethod
    fun updateName(spanBridgeId: String, name: String) {
        val span = getSpan(spanBridgeId) ?: return
        span.updateName(name)
    }

    @ReactMethod
    fun endSpan(spanBridgeId: String, endTime: Double) {
        val span = getSpan(spanBridgeId) ?: return

        if (endTime == 0.0) {
            span.end()
        } else {
            span.end(endTime.toLong(), TimeUnit.MILLISECONDS)
        }

        activeSpans.remove(spanBridgeId)
        if (completedSpans.size > MAX_STORED_SPANS) {
            log.warning("too many completed spans being tracked, ignoring")
        } else {
            completedSpans[spanBridgeId] = span
        }
    }

    @ReactMethod
    fun clearCompletedSpans() {
        completedSpans.clear()
    }
}
